lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp

```diff
@@ -532,6 +537,151 @@ class ConvertTritonGPUToLLVM
+    auto newEpliogueCvt = builder.create<triton::gpu::ConvertLayoutOp>(
+        cvtOp.getLoc(), newDstType, tmpCvt);
+
+    return std::make_pair(tmpCvt, newEpliogueCvt);
+  }
+
+  // Try to reduce LDS usage of cvt(mfma->blocked) op by changing the shape of
+  // WarpsPerCta attribute in mfma layout. The implicit LDS usage of
+  // cvt(mfma->blocked) op depends on the number of warps per CTA that mfma
+  // layout uses along x dimension and block layout uses across y dimension.
+  //
+  // clang-format off
+  //
+  // LDS usage of this op is roughly calculated as:
+  // LDS_USAGE = getShapePerCTA(mfma_layout)[0] * getShapePerCTA(blocked_layoput)[1] * sizeof(data_type)
+  // LDS_USAGE = warpsPerCTA(mfma_layout)[0] * warpsPerCta(blocked_layout)[1] * C,
+  // where C = 32 * sizePerWarp(blocked_layout)[1] * threadsPerWarp(blocked_layout)[1] * sizeof(data_type)
+  //
+  // clang-format on
+  //
+  // When LDS_USAGE exceeds the size of LDS, try to lower LDS usage by
+  // decomposing cvt(mfma->blocked) op into 2 conversions: cvt(mfma->mfma_tmp)
+  // and cvt(mfma_tmp->blocked), where mfma_tmp has WarpsPerCta attribute that
+  // minimizes uses of LDS for these conversions.
+  void reduceCvtOpLDSUsage(ModuleOp mod) const {
+    mod.walk([&](triton::gpu::ConvertLayoutOp cvtOp) -> void {
+      OpBuilder builder(cvtOp);
+
+      auto srcType = cvtOp.getOperand().getType().cast<RankedTensorType>();
+      auto dstType = cvtOp.getType().cast<RankedTensorType>();
+
+      auto srcMfma =
+          srcType.getEncoding().dyn_cast<triton::gpu::MfmaEncodingAttr>();
+      auto dstBlocked =
+          dstType.getEncoding().dyn_cast<triton::gpu::BlockedEncodingAttr>();
+
+      if (!srcMfma || !dstBlocked) {
+        return;
+      }
+
+      auto currLDSUsage = getCvtOpLDSUsage(cvtOp);
+      if (currLDSUsage <= LDSSize) {
+        return;
+      }
+
+      unsigned numWarps =
+          srcMfma.getWarpsPerCTA()[0] * srcMfma.getWarpsPerCTA()[1];
+
+      triton::gpu::ConvertLayoutOp tmpCvt;
+      triton::gpu::ConvertLayoutOp newEpliogueCvt;
+
+      // Find all possible shapes of WarpsPerCTA by finding all possible
+      // factorizations of numWarps. Pick shape for which both conversions in
+      // decomposition use LDS less than LDSSize and for which sum of LDS usage
+      // is minimal. If no such shape exists, do not decompose.
+      unsigned minLDSUsage = 2 * LDSSize;
+      int minIdx = -1;
+      auto factorizedNumWarps = factorizePowerOf2(numWarps);
+
+      for (int i = 0; i < factorizedNumWarps.size(); i++) {
+        auto warpsPerCTAPair = factorizedNumWarps[i];
+        std::tie(tmpCvt, newEpliogueCvt) =
+            createNewConvertOps(mod, builder, cvtOp, warpsPerCTAPair);
+
+        int tmpCvtLDS = getCvtOpLDSUsage(tmpCvt);
+        int newCvtLDS = getCvtOpLDSUsage(newEpliogueCvt);
+        if (tmpCvtLDS <= LDSSize && newCvtLDS <= LDSSize) {
+          int LDSUsage = tmpCvtLDS + newCvtLDS;
```

Comment:
Does lifetimes of scratch buffers of `tmpCvt` and `newEpilogueCvt` overlap?
If not I think it is better to choose maximum from these values instead of sum them.